awx/ui_next/src/components/Workflow/workflowReducer.js

```diff
@@ -0,0 +1,632 @@
+      arrayOfLinksForChart.push({
+        source: arrayOfNodesForChart[sourceIndex],
+        target: arrayOfNodesForChart[targetIndex],
+        linkType: 'failure',
+        type: 'link',
+      });
+      nonRootNodeIds.push(nodeId);
+    });
+    node.always_nodes.forEach(nodeId => {
+      const targetIndex =
+        chartNodeIdToIndexMapping[nodeIdToChartNodeIdMapping[nodeId]];
+      arrayOfLinksForChart.push({
+        source: arrayOfNodesForChart[sourceIndex],
+        target: arrayOfNodesForChart[targetIndex],
+        linkType: 'always',
+        type: 'link',
+      });
+      nonRootNodeIds.push(nodeId);
+    });
+  });
+
+  return [arrayOfLinksForChart, nonRootNodeIds];
+}
+
+// TODO: check to make sure passing i18n into this reducer
+// actually works the way we want it to.  If not we may
+// have to explore other options
+function generateNodesAndLinks(state, workflowNodes, i18n) {
+  const [
+    arrayOfNodesForChart,
+    allNodeIds,
+    nodeIdToChartNodeIdMapping,
+    chartNodeIdToIndexMapping,
+    nodeIdCounter,
+  ] = generateNodes(workflowNodes, i18n);
+  const [arrayOfLinksForChart, nonRootNodeIds] = generateLinks(
+    workflowNodes,
+    chartNodeIdToIndexMapping,
+    nodeIdToChartNodeIdMapping,
+    arrayOfNodesForChart
+  );
+
+  const uniqueNonRootNodeIds = Array.from(new Set(nonRootNodeIds));
+
+  const rootNodes = allNodeIds.filter(
+    nodeId => !uniqueNonRootNodeIds.includes(nodeId)
+  );
+
+  rootNodes.forEach(rootNodeId => {
+    const targetIndex =
+      chartNodeIdToIndexMapping[nodeIdToChartNodeIdMapping[rootNodeId]];
+    arrayOfLinksForChart.push({
+      source: arrayOfNodesForChart[0],
+      target: arrayOfNodesForChart[targetIndex],
+      linkType: 'always',
+      type: 'link',
+    });
+  });
+
+  return {
+    ...state,
+    links: arrayOfLinksForChart,
+    nodes: arrayOfNodesForChart,
+    nextNodeId: nodeIdCounter,
+  };
+}
+
+function selectSourceForLinking(state, sourceNode) {
+  const { links, nodes } = state;
+  const newNodes = [...nodes];
+  const parentMap = {};
+  const invalidLinkTargetIds = [];
+  // Find and mark any ancestors as disabled to prevent cycles
+  links.forEach(link => {
+    // id=1 is our artificial root node so we don't care about that
+    if (link.source.id === 1) {
+      return;
+    }
+    if (link.source.id === sourceNode.id) {
+      // Disables direct children from the add link process
+      invalidLinkTargetIds.push(link.target.id);
+    }
+    if (!parentMap[link.target.id]) {
+      parentMap[link.target.id] = [];
+    }
+    parentMap[link.target.id].push(link.source.id);
+  });
+
+  const getAncestors = id => {
+    if (parentMap[id]) {
+      parentMap[id].forEach(parentId => {
+        invalidLinkTargetIds.push(parentId);
+        getAncestors(parentId);
+      });
+    }
+  };
+
+  getAncestors(sourceNode.id);
+
+  // Filter out the duplicates
+  invalidLinkTargetIds
+    .filter((element, index, array) => index === array.indexOf(element))
```

Comment:
I think this check will always return true, so this `filter` has no effect

Reply:
This filters the duplicate node id's out of the initial array.  

<img width="593" alt="Screen Shot 2020-01-30 at 1 23 25 PM" src="https://user-images.githubusercontent.com/9889020/73478234-ba9e9600-4363-11ea-9a3e-0be83018a4aa.png">

While not 100% necessary this potentially reduces the amount of looping we have to do in the next line.